iroh/src/baomap/mem.rs

```
@@ -0,0 +1,497 @@
+            let mut state = self.state.write().unwrap();
+            let outboard = PreOrderOutboard {
+                root: entry.outboard.root,
+                tree: entry.outboard.tree,
+                data: outboard,
+            };
+            state.partial.remove(&hash);
+            state.complete.insert(hash, (data, outboard));
+            Ok(())
+        }
+        .boxed()
+    }
+}
+
+impl baomap::Store for Store {
+    fn import(
+        &self,
+        data: std::path::PathBuf,
+        _mode: ImportMode,
+        _progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
+    ) -> BoxFuture<'_, io::Result<(Hash, u64)>> {
+        let this = self.clone();
+        tokio::task::spawn_blocking(move || {
+            let bytes: Bytes = std::fs::read(data)?.into();
+            let size = bytes.len() as u64;
+            let hash = this.import_bytes_sync(bytes);
+            Ok((hash, size))
+        })
+        .map(flatten_to_io)
+        .boxed()
+    }
+
+    fn import_bytes(&self, bytes: Bytes) -> BoxFuture<'_, io::Result<Hash>> {
+        let this = self.clone();
+        tokio::task::spawn_blocking(move || {
+            let hash = this.import_bytes_sync(bytes);
+            Ok(hash)
+        })
+        .map(flatten_to_io)
+        .boxed()
+    }
+}
+
+impl Store {
+    fn import_bytes_sync(&self, bytes: Bytes) -> Hash {
+        let size = bytes.len() as u64;
+        let (outboard, hash) = bao_tree::io::outboard(&bytes, IROH_BLOCK_SIZE);
+        let tree = BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE);
+        let outboard = PreOrderOutboard {
+            root: hash,
+            tree,
+            data: outboard.into(),
+        };
+        self.state
+            .write()
+            .unwrap()
+            .complete
+            .insert(hash.into(), (bytes, outboard));
+        hash.into()
+    }
+
+    fn export_sync(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        _mode: ExportMode,
```

why ignored?

---

ExportMode does not really make sense for a mem store. It can not support ExportMode::Reference

---

right, but I think it should error on that, rather than silently ignore it

---

Maybe `TryReference`?

---

So make it a hint, and change the name so that it is clear that is a hint. I can live with that.